### ISSUE TITLE
False positive `invalid-*-returned` when returning an instance of a subclass of the builtin (e.g. `str` subclass returning `self`)

### ISSUE THREAD

**Issue by ekanshul** [PyCQA/pylint#11306]
### Bug description

The `invalid-*-returned` checks only accept an instance whose class is *exactly* the builtin. An instance of a subclass of the builtin is reported, even though returning it is valid (and common: `return self` from a `str` subclass).

The stdlib itself trips it. `pdb._rstr` is a `str` subclass whose `__repr__` returns `self`:

```python
class _rstr(str):
    """String that doesn't quote its repr."""
    def __repr__(self):
        return self
```

Every check that goes through `SpecialMethodsChecker._is_wrapped_type` is affected:

```python
class MyStr(str):
    def __str__(self):  # invalid-str-returned
        return self

    def __repr__(self):  # invalid-repr-returned
        return self

    def __format__(self, spec):  # invalid-format-returned
        return self


class MyInt(int):
    def __hash__(self):  # invalid-hash-returned
        return self

    def __index__(self):  # invalid-index-returned
        return self

    def __len__(self):  # invalid-length-returned
        return self

    def __length_hint__(self):  # invalid-length-hint-returned
        return self


class MyBytes(bytes):
    def __bytes__(self):  # invalid-bytes-returned
        return self


class MyTuple(tuple):
    def __getnewargs__(self):  # invalid-getnewargs-returned
        return self
```

Returning such an instance from an unrelated class (`return MyStr("x")` in some other class's `__str__`) is reported the same way.

### Configuration

```ini

```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
a.py:2:4: E0307: __str__ does not return str (invalid-str-returned)
a.py:5:4: E0306: __repr__ does not return str (invalid-repr-returned)
a.py:8:4: E0311: __format__ does not return str (invalid-format-returned)
a.py:13:4: E0309: __hash__ does not return int (invalid-hash-returned)
a.py:16:4: E0305: __index__ does not return int (invalid-index-returned)
a.py:19:4: E0303: __len__ does not return non-negative integer (invalid-length-returned)
a.py:22:4: E0310: __length_hint__ does not return non-negative integer (invalid-length-hint-returned)
a.py:27:4: E0308: __bytes__ does not return bytes (invalid-bytes-returned)
a.py:32:4: E0312: __getnewargs__ does not return a tuple (invalid-getnewargs-returned)
```

### Expected behavior

No message: an instance of a `str` subclass is a `str`, and CPython accepts it everywhere these protocols are used (`str()`, `repr()`, `format()`, `hash()`, `len()`, `bytes()`, `operator.length_hint()`, pickling).

### Pylint version

```shell
pylint 4.1.0-dev0
astroid 4.2.0b5
Python 3.14.7
```

Also reproduced with pylint 4.0.6.

### OS / Environment

macOS
